cmd/ceremony/main.go

```diff
@@ -380,169 +565,185 @@ func (cc crlConfig) validate() error {
 	if err != nil {
-		return fmt.Errorf("failed to read public key %q: %s", config.Inputs.PublicKeyPath, err)
-	}
-	pubPEM, _ := pem.Decode(pubPEMBytes)
-	if pubPEM == nil {
-		return fmt.Errorf("failed to parse public key")
-	}
-	pub, err := x509.ParsePKIXPublicKey(pubPEM.Bytes)
-	if err != nil {
-		return fmt.Errorf("failed to parse public key: %s", err)
+		return err
 	}
+
 	issuer, err := loadCert(config.Inputs.IssuerCertificatePath)
 	if err != nil {
 		return fmt.Errorf("failed to load issuer certificate %q: %s", config.Inputs.IssuerCertificatePath, err)
 	}
 
+	toBeCrossSigned, err := loadCert(config.Inputs.CertificateToCrossSignPath)
+	if err != nil {
+		return fmt.Errorf("failed to load toBeCrossSigned certificate %q: %s", config.Inputs.CertificateToCrossSignPath, err)
+	}
+
 	signer, randReader, err := openSigner(config.PKCS11, issuer.PublicKey)
 	if err != nil {
 		return err
 	}
 
-	template, err := makeTemplate(randReader, &config.CertProfile, pubPEM.Bytes, ct)
+	template, err := makeTemplate(randReader, &config.CertProfile, pubBytes, ct)
 	if err != nil {
 		return fmt.Errorf("failed to create certificate profile: %s", err)
 	}
 	template.AuthorityKeyId = issuer.SubjectKeyId
 
-	err = signAndWriteCert(template, issuer, pub, signer, config.Outputs.CertificatePath, config.SkipLints)
+	lintCert, err := issueLintCert(template, issuer, pub, signer, config.SkipLints)
 	if err != nil {
 		return err
 	}
 
+	// Ensure that we've configured the correct certificate to cross-sign compared to the profile.
+	//
+	// Example of a misconfiguration below:
+	//      ...
+	//	 	inputs:
+	//  		certificate-to-cross-sign-path: int-e6.cert.pem
+	//		certificate-profile:
+	//  		common-name: (FAKE) E5
+	//  		organization: (FAKE) Let's Encrypt
+	//      ...
+	//
+	if !bytes.Equal(toBeCrossSigned.RawSubject, lintCert.RawSubject) {
+		return fmt.Errorf("mismatch between toBeCrossSigned and lintCert RawSubject DER bytes: \"%x\" != \"%x\"", toBeCrossSigned.RawSubject, lintCert.RawSubject)
+	}
+
+	// Verify that the lintCert (and therefore the eventual finalCert) corresponds to the specified issuer certificate.
+	if !bytes.Equal(issuer.RawSubject, lintCert.RawIssuer) {
+		return fmt.Errorf("mismatch between issuer RawSubject and lintCert RawIssuer DER bytes: \"%x\" != \"%x\"", issuer.RawSubject, lintCert.RawIssuer)
+	}
+
+	finalCert, err := signAndWriteCert(template, issuer, pub, signer, config.Outputs.CertificatePath)
+	if err != nil {
+		return err
+	}
+
+	// Verify that x509.CreateCertificate is deterministic and produced
+	// identical DER bytes between the lintCert and finalCert signing
+	// operations. If this fails it's mississuance, but it's better to know
+	// about the problem sooner than later.
+	if !bytes.Equal(lintCert.RawTBSCertificate, finalCert.RawTBSCertificate) {
+		return fmt.Errorf("mismatch between lintCert and finalCert RawTBSCertificate DER bytes: \"%x\" != \"%x\"", lintCert.RawTBSCertificate, finalCert.RawTBSCertificate)
```

Comment:
Any particular reason you do this in the cross-ceremony but not the other ceremonies? The relationship between lint and final should be the same for all ceremony types. (That said: I don't think this check is _particularly_ necessary for any of the ceremony types.)

Reply:
This came out of a conversation with @jsha a few weeks ago.